util/retry/retry.go

```diff
@@ -0,0 +1,291 @@
+func LinearBackoff(increment time.Duration) BackoffStrategy {
+	return func(attempt int) time.Duration {
+		if attempt == 0 {
+			return 0
+		}
+		return time.Duration(attempt) * increment
+	}
+}
+
+// FixedBackoff returns a fixed backoff strategy
+func FixedBackoff(duration time.Duration) BackoffStrategy {
+	return func(attempt int) time.Duration {
+		if attempt == 0 {
+			return 0
+		}
+		return duration
+	}
+}
+
+// NoBackoff returns immediately without waiting
+func NoBackoff() BackoffStrategy {
+	return func(attempt int) time.Duration {
+		return 0
+	}
+}
+
+// ExecuteSync executes a function synchronously with retry logic
+// It respects context cancellation and timeout
+func ExecuteSync(
+	ctx context.Context,
+	fn func() error,
+	opts ...SyncOptions,
+) error {
+	config := NewRetryConfig()
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	return retryLoop(ctx, fn, config, nil)
+}
+
+// ExecuteAsync executes a function asynchronously with retry logic
+// It respects context cancellation and timeout
+// onSuccess and onFailure callbacks will be called exactly once
+func ExecuteAsync(
+	ctx context.Context,
+	fn func() error,
+	onSuccess func(),
+	onFailure func(error),
+	opts ...AsyncOptions,
+) {
+	config := NewRetryConfig()
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	var once sync.Once
+	callSuccess := func() {
+		once.Do(func() {
+			if onSuccess != nil {
+				onSuccess()
+			}
+		})
+	}
+	callFailure := func(err error) {
+		once.Do(func() {
+			if onFailure != nil {
+				onFailure(err)
+			}
+		})
+	}
+
+	go func() {
+		err := retryLoop(ctx, fn, config, nil)
+		if err == nil {
+			callSuccess()
+		} else {
+			callFailure(err)
+		}
+	}()
+}
+
+// ExecuteSyncWithPredicate executes with a predicate to determine if retry should occur
+func ExecuteSyncWithPredicate(
+	ctx context.Context,
+	fn func() error,
+	shouldRetry IsRetryable,
+	opts ...SyncOptions,
+) error {
+	if shouldRetry == nil {
+		shouldRetry = RetryableError
+	}
+
+	config := NewRetryConfig()
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	return retryLoop(ctx, fn, config, shouldRetry)
+}
+
+// IsRetryable is a helper to determine if an error should trigger a retry
+// You can implement custom logic based on error types or values
+type IsRetryable func(error) bool
+
+// RetryableError checks if an error is transient and should be retried
+// Common transient errors: temporary network failures, timeouts, 5xx HTTP responses
+func RetryableError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// Example: check for temporary interface
+	if temp, ok := err.(interface{ Temporary() bool }); ok {
+		return temp.Temporary()
+	}
+
+	// Example: check for timeout interface
+	if timeout, ok := err.(interface{ Timeout() bool }); ok {
+		return timeout.Timeout()
+	}
+
+	return false
+}
+
+// retryLoop is the core retry logic shared by all retry functions
+func retryLoop(
+	ctx context.Context,
+	fn func() error,
+	config *Config,
+	shouldRetry IsRetryable,
```

Comment:
remove it. This looks meaningless. Your function says `retryLoop` then here you set it to False. Foe what?